tests/NLog.UnitTests/Layouts/LayoutParserTests.cs

```diff
@@ -0,0 +1,98 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using NLog.LayoutRenderers;
+using NLog.LayoutRenderers.Wrappers;
+using NLog.Layouts;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace NLog.UnitTests.Layouts
+{
+    public class LayoutParserTests : NLogTestBase
+    {
+        /// <summary>
+        /// https://github.com/NLog/NLog/issues/3193
+        /// </summary>
+        /// <param name="input"></param>
+        [Theory]
+        [InlineData(@"                                    ${appdomain:format={0\} {1\}}")]
+        [InlineData(@"                           ${cached:${appdomain:format={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${appdomain:format={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${appdomain:format={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${appdomain:format={0\} {1\}}}}}}")]
+        [InlineData(@"                                    ${literal:text={0\} {1\}}")]
+        [InlineData(@"                           ${cached:${literal:text={0\} {1\}}}")]
+        [InlineData(@"                  ${cached:${cached:${literal:text={0\} {1\}}}}")]
+        [InlineData(@"         ${cached:${cached:${cached:${literal:text={0\} {1\}}}}}")]
+        [InlineData(@"${cached:${cached:${cached:${cached:${literal:text={0\} {1\}}}}}}")]
+        public void Issue_3193_Nested_Сlosing_Braces(string input)
+        {
+            var reader = new NLog.Internal.SimpleStringReader(input.Trim());
+            var factory = NLog.Config.ConfigurationItemFactory.Default;
+            IReadOnlyCollection<LayoutRenderer> renderers = LayoutParser.CompileLayout(factory, reader, isNested: true, text: out var text);
+
+            while (true)
+            {
+                Assert.Equal(1, renderers.Count);
+                var singleRender = renderers.Single();
+                if (singleRender is WrapperLayoutRendererBase wrapperRender)
+                {
+                    renderers = ((SimpleLayout)wrapperRender.Inner).Renderers;
+                }
+                else
+                {
+                    break;
+                }
+            }
+            var textRenderer = (LiteralLayoutRenderer)renderers.Single();
+            if (input.IndexOf("literal") >= 0)
+            {
+                Assert.Equal("{0} {1}", textRenderer.Text);
+            }
+            else if (input.IndexOf("appdomain") >= 0)
+            {
+                var appDomain = LogFactory.CurrentAppDomain;
+                Assert.Equal($"{appDomain.Id} {appDomain.FriendlyName}", textRenderer.Text);
+            }
+            else
+            {
+                throw new Xunit.Sdk.XunitException("NOT SUPPORTED");
```

Comment:
Is this assert.Fail?

Reply:
For NUnit - yes
```NUnit.Framework.Assert.Fail("..")```
https://stackoverflow.com/questions/14631923/cannot-find-assert-fail-and-assert-pass-or-equivalent

I updated test.
I think, whether to move the code in ```SimpleLayoutTests.cs```.